File: backend/app/routers/workflows.py
"""
Workflow Automation / Automatisierte Praxis-Abläufe
===================================================
Automatisiert wiederkehrende Aufgaben und Benachrichtigungen.
"""

from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlalchemy.ext.asyncio import AsyncSession
from datetime import datetime, date, timedelta, timezone
from typing import Optional, Any
from enum import Enum
import uuid
import logging

from ..database import get_db
from ..dependencies import get_current_user
from ..models.models import User
from pydantic import BaseModel

router = APIRouter(prefix="/workflows", tags=["Workflow Automation"])
logger = logging.getLogger(__name__)


# =============================================================================
# SCHEMAS
# =============================================================================

class TriggerType(str, Enum):
    """Auslöser für Workflows"""
    appointment_booked = "appointment_booked"
    appointment_reminder = "appointment_reminder"
    appointment_missed = "appointment_missed"
    lab_result_ready = "lab_result_ready"
    prescription_due = "prescription_due"
    vaccination_due = "vaccination_due"
    checkup_reminder = "checkup_reminder"
    document_ready = "document_ready"
    birthday = "birthday"
    manual = "manual"


class ActionType(str, Enum):
    """Aktionen die ausgeführt werden können"""
    send_email = "send_email"
    send_sms = "send_sms"
    send_push = "send_push"
    create_task = "create_task"
    update_record = "update_record"
    generate_document = "generate_document"
    webhook = "webhook"


class WorkflowStatus(str, Enum):
    """Status eines Workflows"""
    active = "active"
    paused = "paused"
    disabled = "disabled"


class WorkflowAction(BaseModel):
    """Einzelne Aktion in einem Workflow"""
    id: str
    type: ActionType
    config: dict
    delay_minutes: int = 0
    condition: Optional[str] = None


class Workflow(BaseModel):
    """Workflow-Definition"""
    id: str
    name: str
    description: str
    trigger: TriggerType
    trigger_config: dict[str, Any] = {}
    actions: list[WorkflowAction]
    status: WorkflowStatus = WorkflowStatus.active
    created_at: datetime
    updated_at: datetime
    run_count: int = 0
    last_run: Optional[datetime] = None


class WorkflowRun(BaseModel):
    """Einzelne Workflow-Ausführung"""
    id: str
    workflow_id: str
    trigger_data: dict
    started_at: datetime
    completed_at: Optional[datetime] = None
    status: str  # running, completed, failed
    actions_completed: int = 0
    error: Optional[str] = None


class TaskPriority(str, Enum):
    """Aufgaben-Priorität"""
    low = "low"
    medium = "medium"
    high = "high"
    urgent = "urgent"


class PracticeTask(BaseModel):
    """Praxis-Aufgabe (automatisch oder manuell erstellt)"""
    id: str
    title: str
    description: Optional[str] = None
    patient_id: Optional[str] = None
    assigned_to: Optional[str] = None
    priority: TaskPriority = TaskPriority.medium
    due_date: Optional[date] = None
    completed: bool = False
    completed_at: Optional[datetime] = None
    created_by: str
    created_at: datetime
    workflow_run_id: Optional[str] = None  # Falls durch Workflow erstellt


# =============================================================================
# VORDEFINIERTE WORKFLOWS
# =============================================================================

PREDEFINED_WORKFLOWS = [
    Workflow(
        id="wf-appointment-reminder-24h",
        name="Termin-Erinnerung (24h vorher)",
        description="Sendet automatisch Push/E-Mail 24 Stunden vor dem Termin",
        trigger=TriggerType.appointment_reminder,
        trigger_config={"hours_before": 24},
        actions=[
            WorkflowAction(
                id="a1",
                type=ActionType.send_push,
                config={
                    "title": "Termin-Erinnerung",
                    "body": "Ihr Termin morgen um {time} Uhr",
                },
            ),
            WorkflowAction(
                id="a2",
                type=ActionType.send_email,
                config={
                    "template": "appointment_reminder",
                },
            ),
        ],
        status=WorkflowStatus.active,
        created_at=datetime(2024, 1, 1),
        updated_at=datetime(2024, 1, 1),
        run_count=156,
    ),
    Workflow(
        id="wf-appointment-reminder-2h",
        name="Termin-Erinnerung (2h vorher)",
        description="Kurze Push-Erinnerung 2 Stunden vor dem Termin",
        trigger=TriggerType.appointment_reminder,
        trigger_config={"hours_before": 2},
        actions=[
            WorkflowAction(
                id="a1",
                type=ActionType.send_push,
                config={
                    "title": "Bald Termin!",
                    "body": "Ihr Termin beginnt in 2 Stunden",
                },
            ),
        ],
        status=WorkflowStatus.active,
        created_at=datetime(2024, 1, 1),
        updated_at=datetime(2024, 1, 1),
        run_count=142,
    ),
    Workflow(
        id="wf-lab-result-notification",
        name="Befund-Benachrichtigung",
        description="Informiert Patienten wenn Laborbefunde freigegeben werden",
        trigger=TriggerType.lab_result_ready,
        actions=[
            WorkflowAction(
                id="a1",
                type=ActionType.send_push,
                config={
                    "title": "Neue Laborergebnisse",
                    "body": "Ihre Laborergebnisse sind jetzt verfügbar",
                },
            ),
            WorkflowAction(
                id="a2",
                type=ActionType.send_email,
                config={
                    "template": "lab_result_ready",
                },
            ),
        ],
        status=WorkflowStatus.active,
        created_at=datetime(2024, 1, 1),
        updated_at=datetime(2024, 1, 1),
        run_count=45,
    ),
    Workflow(
        id="wf-vaccination-recall",
        name="Impf-Recall",
        description="Erinnert an fällige Impfungen (STIKO-Empfehlungen)",
        trigger=TriggerType.vaccination_due,
        trigger_config={"days_before": 14},
        actions=[
            WorkflowAction(
                id="a1",
                type=ActionType.send_push,
                config={
                    "title": "Impf-Erinnerung",
                    "body": "Ihre {vaccine_name}-Impfung ist bald fällig",
                },
            ),
            WorkflowAction(
                id="a2",
                type=ActionType.create_task,
                config={
                    "title": "Impf-Recall: {patient_name}",
                    "priority": "medium",
                },
                delay_minutes=1440,  # 24h später Task für MFA
            ),
        ],
        status=WorkflowStatus.active,
        created_at=datetime(2024, 1, 1),
        updated_at=datetime(2024, 1, 1),
        run_count=28,
    ),
    Workflow(
        id="wf-checkup-reminder",
        name="Vorsorge-Erinnerung (jährlich)",
        description="Erinnert an jährliche Vorsorgeuntersuchung",
        trigger=TriggerType.checkup_reminder,
        trigger_config={"months_since_last": 12},
        actions=[
            WorkflowAction(
                id="a1",
                type=ActionType.send_email,
                config={
                    "template": "checkup_reminder",
                },
            ),
            WorkflowAction(
                id="a2",
                type=ActionType.send_push,
                config={
                    "title": "Zeit für Vorsorge",
                    "body": "Ihre letzte Vorsorge ist über ein Jahr her",
                },
            ),
        ],
        status=WorkflowStatus.active,
        created_at=datetime(2024, 1, 1),
        updated_at=datetime(2024, 1, 1),
        run_count=89,
    ),
    Workflow(
        id="wf-no-show-followup",
        name="Nicht-Erschienen Nachverfolgung",
        description="Erstellt Aufgabe wenn Patient nicht zum Termin erscheint",
        trigger=TriggerType.appointment_missed,
        actions=[
            WorkflowAction(
                id="a1",
                type=ActionType.create_task,
                config={
                    "title": "Anruf: Patient nicht erschienen",
                    "description": "Patient {patient_name} nicht zum Termin am {date} erschienen. Bitte nachfassen.",
                    "priority": "high",
                },
            ),
            WorkflowAction(
                id="a2",
                type=ActionType.send_sms,
                config={
                    "template": "missed_appointment",
                },
                delay_minutes=60,  # 1h später SMS
            ),
        ],
        status=WorkflowStatus.active,
        created_at=datetime(2024, 1, 1),
        updated_at=datetime(2024, 1, 1),
        run_count=12,
    ),
    Workflow(
        id="wf-prescription-refill",
        name="Rezept-Nachbestellung",
        description="Erinnert wenn Dauermedikation bald aufgebraucht",
        trigger=TriggerType.prescription_due,
        trigger_config={"days_before": 7},
        actions=[
            WorkflowAction(
                id="a1",
                type=ActionType.send_push,
                config={
                    "title": "Rezept bald nötig",
                    "body": "Ihr Medikament {medication_name} ist bald aufgebraucht",
                },
            ),
        ],
        status=WorkflowStatus.active,
        created_at=datetime(2024, 1, 1),
        updated_at=datetime(2024, 1, 1),
        run_count=67,
    ),
]

_workflows: dict[str, Workflow] = {w.id: w for w in PREDEFINED_WORKFLOWS}
_tasks: dict[str, PracticeTask] = {}
_runs: dict[str, WorkflowRun] = {}


# =============================================================================
# WORKFLOW ENDPOINTS
# =============================================================================

@router.get("/", response_model=list[Workflow])
async def list_workflows(
    status: Optional[WorkflowStatus] = None,
    trigger: Optional[TriggerType] = None,
):
    """
    Listet alle konfigurierten Workflows.
    """
    workflows = list(_workflows.values())
    
    if status:
        workflows = [w for w in workflows if w.status == status]
    if trigger:
        workflows = [w for w in workflows if w.trigger == trigger]
    
    return workflows


@router.get("/{workflow_id}", response_model=Workflow)
async def get_workflow(workflow_id: str):
    """Holt Details eines Workflows."""
    if workflow_id not in _workflows:
        raise HTTPException(status_code=404, detail="Workflow not found")
    return _workflows[workflow_id]


@router.put("/{workflow_id}/status")
async def update_workflow_status(
    workflow_id: str,
    status: WorkflowStatus,
    current_user: User = Depends(get_current_user),
):
    """Aktiviert/Deaktiviert einen Workflow."""
    if workflow_id not in _workflows:
        raise HTTPException(status_code=404, detail="Workflow not found")
    
    workflow = _workflows[workflow_id]
    workflow.status = status
    workflow.updated_at = datetime.now(timezone.utc)
    
    return {"status": "updated", "new_status": status}


@router.post("/{workflow_id}/trigger")
async def manual_trigger_workflow(
    workflow_id: str,
    trigger_data: dict[str, Any],
    background_tasks: BackgroundTasks,
    current_user: User = Depends(get_current_user),
):
    """
    Löst einen Workflow manuell aus (für Tests).
    """
    if workflow_id not in _workflows:
        raise HTTPException(status_code=404, detail="Workflow not found")
    
    workflow = _workflows[workflow_id]
    
    run_id = str(uuid.uuid4())
    run = WorkflowRun(
        id=run_id,
        workflow_id=workflow_id,
        trigger_data=trigger_data,
        started_at=datetime.now(timezone.utc),
        status="running",
    )
    _runs[run_id] = run
    
    # Execute in background
    background_tasks.add_task(_execute_workflow, workflow, run, trigger_data)
    
    return {"run_id": run_id, "status": "triggered"}


async def _execute_workflow(workflow: Workflow, run: WorkflowRun, data: dict[str, Any]):
    """Führt Workflow-Aktionen aus."""
    try:
        for action in workflow.actions:
            logger.info(f"Executing action {action.id}: {action.type}")
            
            # TODO: Echte Aktionen implementieren
            # await _execute_action(action, data)
            
            run.actions_completed += 1
        
        run.status = "completed"
        run.completed_at = datetime.now(timezone.utc)
        
        workflow.run_count += 1
        workflow.last_run = datetime.now(timezone.utc)
        
    except Exception as e:
        logger.error(f"Workflow {workflow.id} failed: {e}")
        run.status = "failed"
        run.error = str(e)


# =============================================================================
# TASK ENDPOINTS
# =============================================================================

@router.get("/tasks/", response_model=list[PracticeTask])
async def list_tasks(
    assigned_to: Optional[str] = None,
    completed: Optional[bool] = None,
    priority: Optional[TaskPriority] = None,
    current_user: User = Depends(get_current_user),
):
    """
    Listet Praxis-Aufgaben.
    """
    tasks = list(_tasks.values())
    
    if assigned_to:
        tasks = [t for t in tasks if t.assigned_to == assigned_to]
    if completed is not None:
        tasks = [t for t in tasks if t.completed == completed]
    if priority:
        tasks = [t for t in tasks if t.priority == priority]
    
    return sorted(tasks, key=lambda x: (x.completed, x.due_date or date.max))


@router.post("/tasks/", response_model=PracticeTask)
async def create_task(
    title: str,
    description: Optional[str] = None,
    patient_id: Optional[str] = None,
    assigned_to: Optional[str] = None,
    priority: TaskPriority = TaskPriority.medium,
    due_date: Optional[date] = None,
    current_user: User = Depends(get_current_user),
):
    """Erstellt eine neue Aufgabe."""
    task_id = str(uuid.uuid4())
    
    task = PracticeTask(
        id=task_id,
        title=title,
        description=description,
        patient_id=patient_id,
        assigned_to=assigned_to,
        priority=priority,
        due_date=due_date,
        created_by=str(current_user.id),
        created_at=datetime.now(timezone.utc),
    )
    
    _tasks[task_id] = task
    return task


@router.put("/tasks/{task_id}/complete")
async def complete_task(
    task_id: str,
    current_user: User = Depends(get_current_user),
):
    """Markiert eine Aufgabe als erledigt."""
    if task_id not in _tasks:
        raise HTTPException(status_code=404, detail="Task not found")
    
    task = _tasks[task_id]
    task.completed = True
    task.completed_at = datetime.now(timezone.utc)
    
    return {"status": "completed"}


# =============================================================================
# SCHEDULER ENDPOINTS (für Cron-Jobs)
# =============================================================================

@router.post("/scheduler/run-due")
async def run_due_workflows(
    background_tasks: BackgroundTasks,
    current_user: User = Depends(get_current_user),
):
    """
    [INTERNAL] Führt alle fälligen automatischen Workflows aus.
    Sollte von einem Scheduler (Cron) aufgerufen werden.
    """
    triggered = []
    
    # Beispiel: Termin-Erinnerungen prüfen
    # TODO: Echte Termin-Daten aus DB holen
    
    return {
        "checked_at": datetime.now(timezone.utc),
        "workflows_triggered": len(triggered),
        "details": triggered,
    }

File: packages/core/lib/src/models/workflows.dart
/// Models for workflow automation.

enum TriggerType {
  appointmentBooked,
  appointmentReminder,
  appointmentMissed,
  labResultReady,
  prescriptionDue,
  vaccinationDue,
  checkupReminder,
  documentReady,
  birthday,
  manual,
}

TriggerType triggerTypeFromJson(String value) {
  switch (value) {
    case 'appointment_booked':
      return TriggerType.appointmentBooked;
    case 'appointment_reminder':
      return TriggerType.appointmentReminder;
    case 'appointment_missed':
      return TriggerType.appointmentMissed;
    case 'lab_result_ready':
      return TriggerType.labResultReady;
    case 'prescription_due':
      return TriggerType.prescriptionDue;
    case 'vaccination_due':
      return TriggerType.vaccinationDue;
    case 'checkup_reminder':
      return TriggerType.checkupReminder;
    case 'document_ready':
      return TriggerType.documentReady;
    case 'birthday':
      return TriggerType.birthday;
    case 'manual':
      return TriggerType.manual;
    default:
      throw ArgumentError('Unknown TriggerType: $value');
  }
}

String triggerTypeToJson(TriggerType type) {
  switch (type) {
    case TriggerType.appointmentBooked:
      return 'appointment_booked';
    case TriggerType.appointmentReminder:
      return 'appointment_reminder';
    case TriggerType.appointmentMissed:
      return 'appointment_missed';
    case TriggerType.labResultReady:
      return 'lab_result_ready';
    case TriggerType.prescriptionDue:
      return 'prescription_due';
    case TriggerType.vaccinationDue:
      return 'vaccination_due';
    case TriggerType.checkupReminder:
      return 'checkup_reminder';
    case TriggerType.documentReady:
      return 'document_ready';
    case TriggerType.birthday:
      return 'birthday';
    case TriggerType.manual:
      return 'manual';
  }
}

enum ActionType {
  sendEmail,
  sendSms,
  sendPush,
  createTask,
  updateRecord,
  generateDocument,
  webhook,
}

ActionType actionTypeFromJson(String value) {
  switch (value) {
    case 'send_email':
      return ActionType.sendEmail;
    case 'send_sms':
      return ActionType.sendSms;
    case 'send_push':
      return ActionType.sendPush;
    case 'create_task':
      return ActionType.createTask;
    case 'update_record':
      return ActionType.updateRecord;
    case 'generate_document':
      return ActionType.generateDocument;
    case 'webhook':
      return ActionType.webhook;
    default:
      throw ArgumentError('Unknown ActionType: $value');
  }
}

enum WorkflowStatus {
  active,
  paused,
  disabled,
}

WorkflowStatus workflowStatusFromJson(String value) {
  switch (value) {
    case 'active':
      return WorkflowStatus.active;
    case 'paused':
      return WorkflowStatus.paused;
    case 'disabled':
      return WorkflowStatus.disabled;
    default:
      throw ArgumentError('Unknown WorkflowStatus: $value');
  }
}

String workflowStatusToJson(WorkflowStatus status) {
  switch (status) {
    case WorkflowStatus.active:
      return 'active';
    case WorkflowStatus.paused:
      return 'paused';
    case WorkflowStatus.disabled:
      return 'disabled';
  }
}

class WorkflowAction {
  final String id;
  final ActionType type;
  final Map<String, dynamic> config;
  final int delayMinutes;
  final String? condition;

  const WorkflowAction({
    required this.id,
    required this.type,
    required this.config,
    required this.delayMinutes,
    this.condition,
  });

  factory WorkflowAction.fromJson(Map<String, dynamic> json) {
    return WorkflowAction(
      id: json['id'] as String,
      type: actionTypeFromJson(json['type'] as String),
      config: Map<String, dynamic>.from(json['config'] as Map),
      delayMinutes: json['delay_minutes'] as int? ?? 0,
      condition: json['condition'] as String?,
    );
  }
}

class Workflow {
  final String id;
  final String name;
  final String description;
  final TriggerType trigger;
  final Map<String, dynamic> triggerConfig;
  final List<WorkflowAction> actions;
  final WorkflowStatus status;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int runCount;
  final DateTime? lastRun;

  const Workflow({
    required this.id,
    required this.name,
    required this.description,
    required this.trigger,
    required this.triggerConfig,
    required this.actions,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    required this.runCount,
    this.lastRun,
  });

  factory Workflow.fromJson(Map<String, dynamic> json) {
    final actions = (json['actions'] as List<dynamic>? ?? [])
        .whereType<Map<String, dynamic>>()
        .map(WorkflowAction.fromJson)
        .toList();

    return Workflow(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String,
      trigger: triggerTypeFromJson(json['trigger'] as String),
      triggerConfig: Map<String, dynamic>.from(
        (json['trigger_config'] as Map?) ?? {},
      ),
      actions: actions,
      status: workflowStatusFromJson(json['status'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      runCount: json['run_count'] as int? ?? 0,
      lastRun: json['last_run'] == null
          ? null
          : DateTime.parse(json['last_run'] as String),
    );
  }
}

class WorkflowTriggerResponse {
  final String runId;
  final String status;

  const WorkflowTriggerResponse({
    required this.runId,
    required this.status,
  });

  factory WorkflowTriggerResponse.fromJson(Map<String, dynamic> json) {
    return WorkflowTriggerResponse(
      runId: json['run_id'] as String,
      status: json['status'] as String,
    );
  }
}

class WorkflowStatusUpdate {
  final String status;
  final String newStatus;

  const WorkflowStatusUpdate({
    required this.status,
    required this.newStatus,
  });

  factory WorkflowStatusUpdate.fromJson(Map<String, dynamic> json) {
    return WorkflowStatusUpdate(
      status: json['status'] as String,
      newStatus: json['new_status'].toString(),
    );
  }
}

enum WorkflowTaskPriority {
  low,
  medium,
  high,
  urgent,
}

WorkflowTaskPriority workflowTaskPriorityFromJson(String value) {
  switch (value) {
    case 'low':
      return WorkflowTaskPriority.low;
    case 'medium':
      return WorkflowTaskPriority.medium;
    case 'high':
      return WorkflowTaskPriority.high;
    case 'urgent':
      return WorkflowTaskPriority.urgent;
    default:
      throw ArgumentError('Unknown WorkflowTaskPriority: $value');
  }
}

String workflowTaskPriorityToJson(WorkflowTaskPriority priority) {
  switch (priority) {
    case WorkflowTaskPriority.low:
      return 'low';
    case WorkflowTaskPriority.medium:
      return 'medium';
    case WorkflowTaskPriority.high:
      return 'high';
    case WorkflowTaskPriority.urgent:
      return 'urgent';
  }
}

class PracticeTask {
  final String id;
  final String title;
  final String? description;
  final String? patientId;
  final String? assignedTo;
  final WorkflowTaskPriority priority;
  final DateTime? dueDate;
  final bool completed;
  final DateTime? completedAt;
  final String createdBy;
  final DateTime createdAt;
  final String? workflowRunId;

  const PracticeTask({
    required this.id,
    required this.title,
    required this.priority,
    required this.completed,
    required this.createdBy,
    required this.createdAt,
    this.description,
    this.patientId,
    this.assignedTo,
    this.dueDate,
    this.completedAt,
    this.workflowRunId,
  });

  factory PracticeTask.fromJson(Map<String, dynamic> json) {
    DateTime? parseDate(dynamic value) {
      if (value == null) return null;
      return DateTime.tryParse(value.toString());
    }

    return PracticeTask(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      patientId: json['patient_id'] as String?,
      assignedTo: json['assigned_to'] as String?,
      priority: workflowTaskPriorityFromJson(json['priority'] as String),
      dueDate: parseDate(json['due_date']),
      completed: json['completed'] as bool? ?? false,
      completedAt: parseDate(json['completed_at']),
      createdBy: json['created_by'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      workflowRunId: json['workflow_run_id'] as String?,
    );
  }
}

class TaskCreateRequest {
  final String title;
  final String? description;
  final String? patientId;
  final String? assignedTo;
  final WorkflowTaskPriority priority;
  final DateTime? dueDate;

  const TaskCreateRequest({
    required this.title,
    this.description,
    this.patientId,
    this.assignedTo,
    this.priority = WorkflowTaskPriority.medium,
    this.dueDate,
  });

  Map<String, dynamic> toJson() => {
        'title': title,
        if (description != null) 'description': description,
        if (patientId != null) 'patient_id': patientId,
        if (assignedTo != null) 'assigned_to': assignedTo,
        'priority': workflowTaskPriorityToJson(priority),
        if (dueDate != null) 'due_date': dueDate!.toIso8601String(),
      };
}

class TaskCompleteResponse {
  final String status;

  const TaskCompleteResponse({required this.status});

  factory TaskCompleteResponse.fromJson(Map<String, dynamic> json) {
    return TaskCompleteResponse(status: json['status'] as String);
  }
}

class WorkflowRunDueResponse {
  final DateTime checkedAt;
  final int workflowsTriggered;
  final List<dynamic> details;

  const WorkflowRunDueResponse({
    required this.checkedAt,
    required this.workflowsTriggered,
    required this.details,
  });

  factory WorkflowRunDueResponse.fromJson(Map<String, dynamic> json) {
    return WorkflowRunDueResponse(
      checkedAt: DateTime.parse(json['checked_at'] as String),
      workflowsTriggered: json['workflows_triggered'] as int? ?? 0,
      details: json['details'] as List<dynamic>? ?? [],
    );
  }
}
